### ISSUE TITLE
Better error message for using outdated moduleResolution: node(10) 

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [microsoft/TypeScript#55104]
# Suggestion

## 🔍 Search Terms

moduleResolution node node10 eol esm commonjs cannot find types

## ✅ Viability Checklist
My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

When a TSConfig is using `"moduleResolution": "node"` or `"node10"` and an import resolves to an npm package whose types declarations that doesn't support Node < 16, the error message in `tsc` is very general:

```plaintext
Cannot find module '...' or its corresponding type declarations.
```

Same in VS Code:

<img width="878" alt="Screenshot in VS Code of 'Cannot find module '@typescript-eslint/utils' or its corresponding type declarations. ts(2307)' in a hover box" src="https://github.com/microsoft/TypeScript/assets/3335181/83768a52-a6a6-431d-ab32-0b955898f2dc">

Could we please have a secondary message added suggesting that the project be switched to using a more modern `"moduleResolution"`? 

This message needs bikeshedding, but a starting proposal:

```plaintext
Cannot find module '...' or its corresponding type declarations.
    Consider upgrading your "moduleResolution" compiler option. See aka.ms/tsconfig#moduleResolution for details.
```

## 📃 Motivating Example

This has been a point of confusion in at least https://github.com/typescript-eslint/typescript-eslint/issues/7284. We've gotten a few duplicate reports. 

## 💻 Use Cases

Same as ☝️, I think. I can't think of anything more specific.

**Comment by RyanCavanaugh** [comment]
People are going to come here and complain if we issue that message and it turns out that the module really isn't there under any setting, and I don't really see a way we could determinatively detect which (if any) other resolution settings would work.

**Comment by JoshuaKGoldberg** [comment]
Hmm. Maybe it can just link to aka.ms/tsconfig#moduleResolution then?

> detect which (if any) other resolution settings would work

Also, I'm realizing now I meant to suggest in the OP that this message should only be there if any other resolution setting would work. 😬

**Comment by RyanCavanaugh** [comment]
Under different module resolution settings, we could even have entirely different source files under analysis, rendering the original import moot. It's not tractable in the general case IMO

**Comment by DanielRosenwasser** [comment]
Could we just have `node10` do the same thing `node16` does and specially handle the error where they would have diverged?

**Comment by DanielRosenwasser** [comment]
Or something more naive where we keep track of whether `exports` was specified anywhere where we checked.

**Comment by blakeembrey** [comment]
@DanielRosenwasser is it possible to use ambiguous language in the presence of “exports”? For example, “This package may be resolvable using (…) because we found type module and exports declared in the package”? 

I’ve only converted a few packages I maintain to be ESM so far, but every time I do I get a ton of issues like https://github.com/blakeembrey/change-case/issues/304. The current message isn’t adequate to show the issue is in user configuration and not the package.

**Comment by DanielRosenwasser** [comment]
We use "may" a lot of the time, so I'm okay with that.

**Comment by bradzacher** [comment]
Coming back to this - this is a pretty persistent pain-point for users of @typescript-eslint.

We've found that a lot of users don't know that:
- `moduleResolution: node` actually means `node10`
- `node16` exists
- `node10` is different to `node16`
- they're using a module resolution that doesn't match their runtime (our minimum supported node version is 16).

We've had to keep [this pinned issue](https://github.com/typescript-eslint/typescript-eslint/issues/7284) up and open to give people docs on how to fix the unclear error message, or else they file an issue with us saying our package is broken.

I'm sure there are a lot of packages out there who continue to include a `types` field in their package.json solely to stave-off the support burden of users who are still (usually erroneously) using `node10` resolution.
We could do this also - but it does seem "smelly" that we would consider changing our config to avoid support burden from an unclear TS error message.

I 110% get that we can't make the error message say something so absolute like "you're using `node10 `but this package doesn't have a `types` field declared, but it has an `exports` field, you must use `node16` instead" because, well, it's a different module resolution and that advice may not even be correct!
But as @blakeembrey suggested maybe there's a way we can help nudge the user in the right direction for this specific case? Or at least suggest something that might help them try something else and self-service fix their problem, rather than having to reach for stackoverflow / github issues / discord messages in order to even learn about the 2 character config change that could fix their problem.

For example, would this rough logic be acceptable?
"if TS finds a package that has no `types` field but has an `exports` field with a nested `types` field, then append ``This package may be resolvable using (…) because we found typed exports declared in the package.`` to the error message"
